### PR TITLE
Updates to capacity management

### PR DESCRIPTION
**Pull request by mike-tutkowski**
## Description
In StorageManagerImpl.storagePoolHasEnoughSpace, we need to update a couple areas of the algorithm that calculates if enough space is present when dealing with managed storage:

1) We no longer can rely on managed storage being exclusively at the zone level. Check if the storage is managed (not if if it at the zone level).

2) Invoke getBytesRequiredForTemplate not only for XenServer when getSupportsResigning resolves to true, but also if using VMware or KVM.

https://issues.apache.org/jira/browse/CLOUDSTACK-10335

## Types of changes
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x ] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

## How Has This Been Tested?

Initially I noticed on VMware and KVM that templates were not being included in the space used for primary storage when that storage is managed. I made the necessary changes (included in this PR) and then checked space used to verify that the new calculated number was now accurate for managed storage when using those hypervisor types.

## Checklist:
- [x ] I have read the [CONTRIBUTING](https://github.com/apache/cloudstack/blob/master/CONTRIBUTING.md) document.
- [ x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ x] All new and existing tests passed.

@blueorangutan package

### CODE REVIEW

**Comment by mike-tutkowski** [comment]
I plan to write a Marvin test for this, but - in the meanwhile - wanted to get this PR opened so reviewers could provide comments on the production-focused code.

**Comment by borisstoyanov** [comment]
@blueorangutan package

**Comment by borisstoyanov** [comment]
@blueorangutan test

**Comment by mike-tutkowski** [comment]
I'm pretty sure none of those test failures has to do with this PR. The PR code relates only to managed storage (which none of those tests test). On top of it, the code is really concerned with somewhat of a corner case in managed storage (which none of those tests would test either).

**Comment by borisstoyanov** [comment]
Yes @mike-tutkowski I think that's absolutely valid. it makes me sad to see these random failures occasionally... :(  
@blueorangutan test

File: server/src/main/java/com/cloud/storage/StorageManagerImpl.java [hunk 1]
```diff
@@ -1833,67 +1839,88 @@ public boolean storagePoolHasEnoughSpace(List<Volume> volumes, StoragePool pool,
...
-                        // Storage plug-ins for zone-wide primary storage can be designed in such a way as to store a template on the
-                        // primary storage once and make use of it in different clusters (via cloning).
+                        // Storage plug-ins for managed storage can be designed in such a way as to store a template on the
+                        // primary storage once and make use of it via storage-side cloning.
                         // This next call leads to CloudStack asking how many more bytes it will need for the template (if the template is
                         // already stored on the primary storage, then the answer is 0).
 
-                        if (clusterId != null && _clusterDao.getSupportsResigning(clusterId)) {
-                            totalAskingSize += getBytesRequiredForTemplate(tmpl, pool);
+                        if (clusterId != null) {
```
Comment (rafaelweingartner) [inline server/src/main/java/com/cloud/storage/StorageManagerImpl.java]: Would you mind extracting the block of this IF condition to a method? This would allow proper documentation an unit tests.

If the resigning is not supported this method can return 0 as the value to be added to the `totalAskingSize` variable.

Comment (mike-tutkowski) [reply server/src/main/java/com/cloud/storage/StorageManagerImpl.java]: Done

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by mike-tutkowski** [comment]
I've added an integration test.

**Comment by mike-tutkowski** [comment]
All test errors seem inapplicable to this PR. Here are some examples:

test_primary_storage.py: errorText:Failed to add data store: Storage pool nfs://10.2.0.16/acs/primary/pr2499-t2464-kvm-centos7/marvin_pri1 already in use by another pod (id=1)\n']

test_snapshots.py: errorText:Failed to add data store: Storage pool nfs://10.2.0.16/acs/primary/pr2499-t2464-kvm-centos7/nfs2 already in use by another pod (id=1)\n']

test_templates.py: 'AssertionError: Extract Template Failed with invalid URL http://192.168.100.96/userdata/99b8334e-ecaa-405b-9168-e902981a3c40.qcow2 (template id: 8cc43b7f-00e7-4250-acbc-53be1de58627)\n']

test_vm_life_cycle.py: errortext : u'Cannot migrate VM, destination host is not in correct state, has status: Up, state: Disabled'}, accountid : u'c600e427-38a5-11e8-a6b6-06db8e010701'}\n"]

test_volumes.py: 'AssertionError: Extract Volume Failed with invalid URL http://192.168.100.96/userdata/c146f89d-12e8-4a34-8087-79e66e110239.qcow2 (vol id: ab60d379-a5d3-471a-b17f-7df204e48e53)\n']

**Comment by mike-tutkowski** [comment]
These test failures do not seem to be related to this PR. Here are some examples of the errors:

**test_primary_storage.py:**
errorText:Primary storage with id 5 cannot be disabled.
errorText:Failed to add data store: Storage pool nfs://10.2.0.16/acs/primary/pr2499-t2470-kvm-centos7/marvin_pri1 already in use by another pod (id=1)\n']

**test_snapshots.py:**
errorText:Failed to add data store: Storage pool nfs://10.2.0.16/acs/primary/pr2499-t2470-kvm-centos7/nfs2 already in use by another pod (id=1)\n']

**test_vm_life_cycle.py:**
errortext : u'Cannot migrate VM, destination host is not in correct state, has status: Up, state: Disabled'}, accountid : u'be964da7-397f-11e8-a179-06965801071a'}\n"]

**Comment by borisstoyanov** [comment]
@blueorangutan test

**Comment by mike-tutkowski** [comment]
The test environment is having an issue when we try to put an NFS-based primary storage in maintenance mode. In test_primary_storage.py, the first error is related to that and then we later see other errors where adding a new primary storage with the same name fails because it's already in use (presumably we were originally going to delete the primary storage after putting it in maintenance mode, but putting it in maintenance mode failed).

Is this error scenario unique to this PR? It seems like the code in this PR wouldn't be responsible for such a situation.

On the up side, both Jenkins and Travis passed.

**test_primary_storage.py:**

errorText:Primary storage with id 5 cannot be disabled. Storage pool state : Maintenance\n'
errorText:Failed to add data store: Storage pool nfs://10.2.0.16/acs/primary/pr2499-t2490-kvm-centos7/marvin_pri1 already in use by another pod (id=1)\n' **(two of these errors)**

**test_routers.py:**

'AssertionError: Check uptime is less than 3 mins or not\n'

**test_snapshots.py:**

errorText:Failed to add data store: Storage pool nfs://10.2.0.16/acs/primary/pr2499-t2490-kvm-centos7/nfs2 already in use by another pod (id=1)\n'

**test_vm_life_cycle.py:**

errortext : u'Cannot migrate VM, destination host is not in correct state, has status: Up, state: Disabled'

**Comment by mike-tutkowski** [comment]
I looked at several of the error messages for a recent test run of #2486 and it seems the list is quite similar to the list of error messages for this PR. As such, I suggest it is likely that none of the errors that are listed for this test run are related to this PR.

**Comment by rhtyd** [comment]
@mike-tutkowski  I scanned PRs merged on master but not 4.11 and came across this PR, do you think this would be useful for 4.11? If so, can you help create a backport PR for 4.11?

**Comment by mike-tutkowski** [comment]
Yes, @rhtyd, you are correct that this PR is a good candidate for 4.11.1. I can create a PR to back port it.

**Comment by rhtyd** [comment]
Thanks @mike-tutkowski, looking forward to your port-PR.

**Comment by mike-tutkowski** [comment]
Hey @rhtyd - What's the official process for backporting a PR like this that has already been merged? Should I just cherry pick the commit to 4.11.1? Clearly there's not much tracking going on that way, but I'm not sure how we officially do this. Thanks!

**Comment by rhtyd** [comment]
@mike-tutkowski yes, please cherry-pick or manuall port your changes to 4.11 branch. You can then either create a new PR, or push to this backport PR https://github.com/apache/cloudstack/pull/2621 (on this PR both rafael and I have push changes)

**Comment by mike-tutkowski** [comment]
I went ahead and pushed the cherry-picked commit to #2621.
